unified_planning/engines/sequential_simulator.py

```diff
@@ -0,0 +1,211 @@
+        lift_map = self._grounding_result.map_back_action_instance
+        assert lift_map is not None
+        self._events: Dict[Tuple['up.model.Action', Tuple['up.model.FNode', ...]], List[Event]] = {}
+        for grounded_action in grounded_problem.actions:
+            if isinstance(grounded_action, up.model.InstantaneousAction):
+                lifted_ai = lift_map(ActionInstance(grounded_action))
+                event_list = self._events.setdefault((lifted_ai.action, lifted_ai.actual_parameters), [])
+                event_list.append(
+                    InstantaneousEvent(
+                        grounded_action.preconditions,
+                        grounded_action.effects,
+                        grounded_action.simulated_effect))
+            else:
+                raise NotImplementedError
+        self._se = StateEvaluator(self._problem)
+
+    def _is_applicable(self, event: 'Event', state: 'up.model.ROState') -> bool:
+        '''
+        Returns True if the given event conditions are evaluated as True in the given state;
+        returns False otherwise.
+        :param state: the state where the event conditions are checked.
+        :param event: the event whose conditions are checked.
+        :return: Whether or not the event is applicable in the given state.
+        '''
+        # Evaluate every condition and if the condition is False or the condition is not simplified as a
+        # boolean constant in the given state, return False. Return True otherwise
+        for c in event.conditions:
+            evaluated_cond = self._se.evaluate(c, state)
+            if not evaluated_cond.is_bool_constant() or not evaluated_cond.bool_constant_value():
+                return False
+        return True
+
+    def _apply(self, event: 'Event', state: 'up.model.RWState') -> Optional['up.model.RWState']:
+        '''
+        Returns None if the event is not applicable in the given state, otherwise returns a new RWState,
+        which is a copy of the given state but the applicable effects of the event are applied; therefore
+        some fluent values are updated.
+        :param state: the state where the event formulas are calculated.
+        :param event: the event that has the information about the conditions to check and the effects to apply.
+        :return: None if the event is not applicable in the given state, a new RWState with some updated values
+         if the event is applicable.
+        '''
+        if not self.is_applicable(event, state):
+            return None
+        else:
+            return self.apply_unsafe(event, state)
+
+    def _apply_unsafe(self, event: 'Event', state: 'up.model.RWState') -> 'up.model.RWState':
+        '''
+        Returns a new RWState, which is a copy of the given state but the applicable effects of the event are applied; therefore
+        some fluent values are updated.
```

Comment:
You are modifying the same state, not a copy of it.
I think you can implement this method using only `ROState`.

Reply:
but the `set_values` method is only defined for RWStates (It does return a new state "copied" from `state`)

Reply:
Where is the state copied?

In any case I think that the input state could be a ROState, since we want only to read from it.

For the returned state, with the current implementation we need it, but in principle we could set the updated values in the constructor of a new ROState.

Reply:
Have a look at the `set_values` method (the name is indeed misleading)

Reply:
Ok, now I got it!

I think both `RWState` and `set_values` are misleading names